peer_heap.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannel
+
+import (
+	"container/heap"
+	"math/rand"
+	"sync/atomic"
+	"time"
+)
+
+// PeerHeap maintains a MIN heap of peers based on the peers' score.
+type PeerHeap struct {
+	PeerScores []*peerScore
+	rng        *rand.Rand
+	order      uint64
+}
+
+func newPeerHeap() *PeerHeap {
+	return &PeerHeap{rng: NewRand(time.Now().UnixNano())}
+}
+
+func (ph PeerHeap) Len() int { return len(ph.PeerScores) }
+
+func (ph *PeerHeap) Less(i, j int) bool {
+	if ph.PeerScores[i].score == ph.PeerScores[j].score {
+		return ph.PeerScores[i].order < ph.PeerScores[j].order
```

Comment:
this introduces harmonics -- if you have 200 go clients, they will all have the same hosts in the same order, and will probably make requests to the same nodes in order.

when you make one request at a time, is this enough to avoid selecting the same nodes repeatedly?